src/Interpreters/GraceHashJoin.cpp

```diff
@@ -555,16 +566,6 @@ IBlocksStreamPtr GraceHashJoin::getDelayedBlocks()
 
     size_t bucket_idx = current_bucket->idx;
 
-    if (hash_join)
-    {
-        auto right_blocks = hash_join->releaseJoinedBlocks(/* restructure */ false);
```

Comment:
We can have rows that belong to different bucket.
For example:
we have right keys [0,2,4] in 0-bucket and [1,3,5] in the 1-bucket and we are reading the right table.
On addJoinedBlock we add keys [6,7] have overflow and do rehash.
Now we have buckets with right keys:
right keys | bucket
--- | ---
[0,2,4] | 0
[1,3,5] | 1
[6]     | 2
[7]     | 3

Then we read left table and add rows to buckets also:
left | right | bucket
--- | --- | ---
[0] | [0,2,4] | 0
[1] | [1,3,5] | 1
[2] | [6]     | 2
[3] | [7]     | 3

Then we go to delayed stage. We read bucket 1 and process keys [1,5] wich belong to it, but row with the key [3] shold go to bucket 3, because we have corresponding left row there that was added after rehash from 2 to 4 buckets. 

Will we handle it in new code in `addJoinedBlockImpl`? Could you explain?

Reply:
There a some rehash operation to avoid this case happen, and the buckets are hadled one after one.

When inserting [6,7] trigger rehash,  bucket[0] will releases all blocks and make a rehash them by the new bucket size, other buckets remain unchanged. So the contents of buckets now are
|  right keys   | bucket  |
|  ----  | ----  |
| [0,4] | 0 |
| [1,3,5]  | 1 |
| [2,4] | 2 |
| [7] | 3 |

When come to  join left blocks(say we have keys [0,1,2,3] in left table) but not in the delayed stage, buckets' size will not be changed, so bucket[0] will be find. What we have after bucket[0]  is
|  left keys   |  right keys   | bucket  |
|  ----  |  ----  | ----  |
| [1] | [1,3,5]  | 1 |
| [2] | [2,4] | 2 |
|[3] | [7] | 3 |

bucket[0]'s right side contains some rows not belong to it, But during loading the right blocks from bucket[0], there also a `scatter` operation. It will deliver key [3] to the bucket[3], now we  have
|  left keys   |  right keys   | bucket  |
|  ----  |  ----  | ----  |
| [1] | [1,5]  | 1 |
| [2] | [2,4] | 2 |
|[3] | [3,7] | 3 |

Then take the  probe stage for bucket[1] left side. Also blocks loading from file will also be `scatter` again, if the bucket[1] left side contains keys of other buckets, they will also be delivered to the new buckets.